### DISCUSSION TITLE
Interim being received after its respective is_final Results

### DISCUSSION THREAD

**Discussion by MattMacGregor** [deepgram/community#941] [General help]
Occasionally I have noticed that for a given piece of audio while streaming, I will received the finalized results (marked with is_final: true), and then later receive a result transcribing the same chunk of audio with is_final: false.

Is this expected behavior, to occasionally receive an interim result after a finalized result when using the Streaming API?

From our logging we see:
At Sep 27 23:17:08, we received Results from Deepgram with is_final and speech_final both true. We then received  some empty transcripts. But after that at Sep 27 23:17:13 we received the same transcript as at Sep 27 23:17:08, but with is_final and speech_final set as false, indicating it as a interim transcription, seemingly coming after its respective finalized transcript. We then received a couple more empty transcripts, followed by another interim results with the transcript “Gwen, what's going”, which is completely different then the one received at 23:17:08 and 23:17:13. This seems to refute any potential issue on our end where audio was sent twice from us, as I would have expected to see that 23:17:13 get finalized again after. Instead it seemed to move on transcribing completely different audio, with that "Gwen, what's going" transcription. 

Request id of the streaming request in question: `42c9d2a8-44b2-4a0f-ada9-2024a8959915`

We are not using any of the SDKs and are using the Websocket API with C++.

Discord post for same issue: https://discord.com/channels/1108042150941294664/1290383402914025533/1290383402914025533

**Comment by jkroll-deepgram** [comment]
Hi @MattMacGregor, do you happen to have those few sequential JSON responses that Deepgram returned?

At a high level, the same transcript should not be returned as a final result, and then again as an interim result.

However, if you're saying that a given transcript is returned as an interim result, but then not as a final result, it may be that Deepgram initially predicted a certain transcript, but then decided not to finalize that transcript, because the interim result was low confidence.

I also see on that request that you're using the keyword `Gwyneth` (though not `Gwen`), which has the potential to increase predictions of that word in the transcript.

Does this sound like the pattern you're seeing?

Final: `"{XYZ}"`
Interim: `"{XYZ}"`
Final: `""`
Interim: `"Gwen, what's going"`
Final: `""`

In that case, the only finalized results would be `{XYZ}`, and the other interim contents were initially predicted by Deepgram, but we did not have high enough confidence to return them as final results.

**Reply by MattMacGregor** [reply]
It's this case:
> At a high level, the same transcript should not be returned as a final result, and then again as an interim result.

For this request we received the following JSON objects (starting with oldest):

```
{"type":"Results","channel_index":[0,1],"duration":1.0,"start":43.22,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"Go on this","confidence":0.37498987,"words":[{"word":"go","start":43.460003,"end":43.620003,"confidence":0.1797258,"punctuated_word":"Go"},{"word":"on","start":43.620003,"end":43.870003,"confidence":0.8796478,"punctuated_word":"on"},{"word":"this","start":43.870003,"end":44.22,"confidence":0.37498987,"punctuated_word":"this"}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":2.0,"start":43.22,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"Gwanis, tell me something.","confidence":0.9987348,"words":[{"word":"gwanis","start":43.620003,"end":44.100002,"confidence":0.81517255,"punctuated_word":"Gwanis,"},{"word":"tell","start":44.100002,"end":44.34,"confidence":0.9987348,"punctuated_word":"tell"},{"word":"me","start":44.34,"end":44.66,"confidence":0.9999387,"punctuated_word":"me"},{"word":"something","start":44.66,"end":45.16,"confidence":0.7409337,"punctuated_word":"something."}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":3.199997,"start":43.22,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"Gwyneth, tell me something new.","confidence":0.9955337,"words":[{"word":"gwyneth","start":43.620003,"end":44.100002,"confidence":0.8051765,"punctuated_word":"Gwyneth,"},{"word":"tell","start":44.100002,"end":44.420002,"confidence":0.9955337,"punctuated_word":"tell"},{"word":"me","start":44.420002,"end":44.66,"confidence":0.999943,"punctuated_word":"me"},{"word":"something","start":44.66,"end":45.16,"confidence":0.9998198,"punctuated_word":"something"},{"word":"new","start":45.460003,"end":45.960003,"confidence":0.98363566,"punctuated_word":"new."}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":1.0,"start":46.42,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":2.0,"start":46.42,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":3.6100006,"start":46.42,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":4.6100006,"start":46.42,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"Gwanis, tell me something new.","confidence":0.8897977,"words":[{"word":"gwanis","start":46.42,"end":46.92,"confidence":0.86089575,"punctuated_word":"Gwanis,"},{"word":"tell","start":47.78,"end":48.28,"confidence":0.8897977,"punctuated_word":"tell"},{"word":"me","start":48.579998,"end":48.739998,"confidence":0.9381839,"punctuated_word":"me"},{"word":"something","start":48.739998,"end":48.98,"confidence":0.8689447,"punctuated_word":"something"},{"word":"new","start":48.98,"end":49.48,"confidence":0.987749,"punctuated_word":"new."}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":3.0200005,"start":46.42,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":1.9799995,"start":49.44,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":2.9799995,"start":49.44,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"Gwen, what's going","confidence":0.97891605,"words":[{"word":"gwen","start":51.039997,"end":51.539997,"confidence":0.6102328,"punctuated_word":"Gwen,"},{"word":"what's","start":51.76,"end":52.0,"confidence":0.97891605,"punctuated_word":"what's"},{"word":"going","start":52.0,"end":52.42,"confidence":0.99652916,"punctuated_word":"going"}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}
```

We get the `is_final` response in the third Result with `"Gwyneth, tell me something new."`, but followed by some empty transcripts. But the 7th Result is `is_final: false` and looks like it transcribed the same audio and resulted in `"Gwanis, tell me something new."`. And it doesn't seem like we sent audio twice because after that 7th JSON, it keeps moving forward and never "finalized" that 7th interim response, which I'd expect to see if we actually just sent the audio twice. As you can see it just starts transcribing `"Gwen, what's going"`, like that 7th interim result never existed.

**Reply by MattMacGregor** [reply]
Additionally the "late interim" with `"Gwanis, tell me something new"` seems to have overallping timestamps with some of the empty transcript messages (all starting at 46.42), which seems odd

**Reply by jkroll-deepgram** [reply]
Hi @MattMacGregor, thanks for providing the JSON responses.

Here's the finalized transcript I'm seeing:

43.22 - 46.42: "Gwyneth, tell me something new." (finalized)
46.42 - 49.44: "" (finalized - ultimately decided that no speech was present, despite the earlier interim predictions for that time span)

I'll repeat this bit from my prior reply, as that's what I believe is happening:
> However, if you're saying that a given transcript is returned as an interim result, but then not as a final result, it may be that Deepgram initially predicted a certain transcript, but then decided not to finalize that transcript, because the interim result was low confidence.

Our models use audio context to predict speech, so my best explanation is that's why there's a similar interim transcript following the final one - the model looked back and thought there may be similar speech ahead, but then in the final transcription, correctly determined that was not the case.

I'm not seeing any overlapping timestamps - the first finalized result ends at 46.42 (i.e., 43.22 + 3.20), which is right when the next one starts.

Zooming out, are you finding that this behavior is causing issues for your application? We may be able to offer guidance on how to use our interim and final results differently so this kind of pattern doesn't impede your application.

**Reply by MattMacGregor** [reply]
While I understand that technically what occurred here with the interim message could be explained by the model changing its mind as it finalized the transcription, it just seems unlikely here, no?

The odds that the audio submitted after the "Gwyneth, tell me something new" audio was sent would happen to result in an interim prediction nearly identical to the previous final result for only a single interim result seems low, especially considering that when that same audio was "finalized" it ended up having no speech. 

If it was a single word, it seems more reasonable, but the model jumping between transcribing a full sentence to no speech at all, seems indicative of something weird going on. And adding in the fact that the full sentence matches the previous audio transcribed, it just seems like a lot of strange occurrences.

Obviously I don't have all the details internally how you buffer and store the audio to be processed, but it really seems like some "old" audio was processed again, and then the new "correct" audio was processed later which correctly was transcribed with no speech.

For context this is an issue because generally speaking we use the INTERIM results as a heads up that someone is speaking, so it coming in with a full sentence followed by "nothing" can screws with the pipeline we use your transcription service with.

Edit: additionally that interim that came in with the "Gwanis, tell me something new." had a confidence of `0.8897977`, which certainly doesn't seem low and makes it seem unlikely that the very next interim the model would predict no speech at all.

**Reply by jkroll-deepgram** [reply]
I think the context windows of Nova-2, plus the use of the keywords feature, raise the probability of the next interim result having stray predictions that incorporate the prior result. Not that it's a bizarre fluke, but that it's a vestige of the model. I haven't seen it happen often myself, but I've seen a handful of cases like this, especially with the use of keywords. Is this one case you noticed, or you're encountering it frequently?

Interim results can be used as a heuristic that someone has started speaking, but they are always subject to change before finalizing. Another complement to detect speech starting is a voice activity detector (VAD). [Silero](https://github.com/snakers4/silero-vad) is an open-source VAD that our customers have found good results with.

**Reply by MattMacGregor** [reply]
It has occurred more than once, but I don't have exact details on the frequency. Enough that these examples were found accidentally just scrolling through the logs searching for another unrelated bug.

Another example from the same day, same request:
```
 {"type":"Results","channel_index":[0,1],"duration":3.4499998,"start":8.47,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"What do you do?","confidence":0.9994704,"words":[{"word":"what","start":10.950001,"end":11.190001,"confidence":0.99823296,"punctuated_word":"What"},{"word":"do","start":11.190001,"end":11.43,"confidence":0.99952686,"punctuated_word":"do"},{"word":"you","start":11.43,"end":11.67,"confidence":0.9994704,"punctuated_word":"you"},{"word":"do","start":11.67,"end":11.92,"confidence":0.768831,"punctuated_word":"do?"}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":4.0299997,"start":8.47,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"What do you do?","confidence":0.99931014,"words":[{"word":"what","start":11.030001,"end":11.35,"confidence":0.9972204,"punctuated_word":"What"},{"word":"do","start":11.35,"end":11.51,"confidence":0.99931014,"punctuated_word":"do"},{"word":"you","start":11.51,"end":11.75,"confidence":0.9996288,"punctuated_word":"you"},{"word":"do","start":11.75,"end":12.25,"confidence":0.9988475,"punctuated_word":"do?"}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":1.0,"start":12.5,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"What do you do?","confidence":0.9986399,"words":[{"word":"what","start":12.5,"end":12.75,"confidence":0.65755695,"punctuated_word":"What"},{"word":"do","start":12.75,"end":13.0,"confidence":0.9986399,"punctuated_word":"do"},{"word":"you","start":13.0,"end":13.25,"confidence":0.99923205,"punctuated_word":"you"},{"word":"do","start":13.25,"end":13.5,"confidence":0.9974319,"punctuated_word":"do?"}]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}

{"type":"Results","channel_index":[0,1],"duration":2.0,"start":12.5,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"42c9d2a8-44b2-4a0f-ada9-2024a8959915","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"},"from_finalize":false}
```

In that example the interim comes after the final with the exact same transcript, since it was the same request there were keywords specified but none in that transcript at least. Additionally the "late interim" there had a confidence of `0.9986399`, which again seems way to high for it to then send a transcript of `""` on the very next interim, without it being some sort of bug.

I'd need to look around a bit more to find other examples, since we don't have any logging to explicitly detect these occurrences, since its inherently a hard thing to detect given we've seen the text is similar but not always exactly the same.

**Reply by jkroll-deepgram** [reply]
@MattMacGregor I see, I understand that this happens on occasion, and it's plausible that it's an artifact of the Nova-2 model, which may be contributing to the high confidence on incorrect answers.

If this is a major blocker to you, you could choose to also test our Enhanced model, which uses context differently, so I'd expect that this issue would not occur with that model. Otherwise, it may be an edge case that you'd handle on occasion.

If it helps, we are also actively developing our new Nova-3 model, which will have its own updated architecture in the next few months, which may mitigate this issue as we're using new strategies on how to incorporate context.
